usr/lib/common/mech_md5.c

```diff
@@ -67,81 +128,73 @@ CK_RV md5_hash(STDLL_TokData_t *tokdata,
         return CKR_BUFFER_TOO_SMALL;
     }
 
-    rc = md5_hash_update(tokdata, sess, ctx, in_data, in_data_len);
-    if (rc != CKR_OK) {
-        TRACE_DEVEL("md5_hash_update failed\n");
-        return rc;
-    }
+    if (ctx->context == NULL)
+        return CKR_HOST_MEMORY;
 
-    return md5_hash_final(tokdata, sess, FALSE, ctx, out_data, out_data_len);
+    if (ctx->mech.mechanism == CKM_MD5)
+        return sw_md5_hash(ctx, in_data, in_data_len, out_data,
+                           out_data_len);
+    else
+        return CKR_MECHANISM_INVALID;
 }
 
-
 //
 //
-CK_RV md5_hash_update(STDLL_TokData_t *tokdata,
-                      SESSION *sess,
-                      DIGEST_CONTEXT *ctx,
-                      CK_BYTE *in_data, CK_ULONG in_data_len)
+CK_RV md5_hash_update(STDLL_TokData_t *tokdata, SESSION *sess,
+                      DIGEST_CONTEXT *ctx, CK_BYTE *in_data,
+                      CK_ULONG in_data_len)
 {
-    if (!sess || !ctx || !in_data) {
-        TRACE_ERROR("%s received bad argument(s)\n", __func__);
-        return CKR_FUNCTION_FAILED;
-    }
+    UNUSED(tokdata);
+    UNUSED(sess);
 
-    return ckm_md5_update(tokdata, (MD5_CONTEXT *) ctx->context, in_data,
-                          in_data_len);
-}
+    /* if no data to hash, just return */
+    if (!in_data_len)
+        return CKR_OK;
 
+    if (ctx->mech.mechanism == CKM_MD5)
+        return sw_MD5_Update(ctx, in_data, in_data_len);
+    else
+        return CKR_MECHANISM_INVALID;
+}
 
-//
-//
-CK_RV md5_hash_final(STDLL_TokData_t *tokdata,
-                     SESSION *sess,
-                     CK_BYTE length_only,
-                     DIGEST_CONTEXT *ctx,
+CK_RV md5_hash_final(STDLL_TokData_t *tokdata, SESSION *sess,
+                     CK_BYTE length_only, DIGEST_CONTEXT *ctx,
                      CK_BYTE *out_data, CK_ULONG *out_data_len)
 {
-    CK_RV rc;
-
+    UNUSED(tokdata);
+    UNUSED(sess);
```

Comment:
looks something is wrong with tabs/spaces

Reply:
The code looks ok in Eclipse: 4 spaces in front of the UNUSED() statements. Maybe a formatting problem here.